File: Graph/860/stain.cpp
#include <iostream>
#include <cstring>
using namespace std;

const int N = 1e5 + 10;
int n, m, x, y, idx, h[N], e[2 * N], color[N], ne[2 * N];
bool flag = true;

void add(int a, int b);
bool stain(int u, int c);

int main() {
    memset(h, -1, sizeof(h));
    scanf("%d%d", &n, &m);
    while (m--) {
        scanf("%d%d", &x, &y);
        add(x, y), add(y, x);
    }
    for (int i = 1; i <= n; i++) {
        if (!color[i]) {
            if (!stain(i, 1)) {
                flag = false;
                break;
            }
        }
    }
    if (flag) printf("Yes");
    else printf("No");
    return 0;
}

void add(int a, int b) {
    e[idx] = b, ne[idx] = h[a], h[a] = idx++;
}

bool stain(int u, int c) {
    color[u] = c;
    for (int i = h[u]; i != -1; i = ne[i]) {
        if (!color[e[i]]) {
            if (!stain(e[i], 3 - c)) return false;
        } else if (color[e[i]] == c) return false;
    }
    return true;
}

File: Graph/860/stain.py
from collections import defaultdict, deque

def bfs(u: int, c: int, g: dict, color: list) -> bool:
    q = deque()
    color[u] = c
    q.append((u, c))
    while q:
        u, c = q.popleft()
        for i in g[u]:
            if not color[i]:
                color[i] = 3 - c
                q.append((i, color[i]))
            elif color[i] == c:
                return False
    return True

if __name__ == "__main__":
    n, m = map(int, input().split())
    g = defaultdict(list)
    color = [0] * (n + 1)
    for _ in range(m):
        x, y = map(int, input().split())
        g[x].append(y)
        g[y].append(x)
    flag = True
    for i in range(1, n + 1):
        if not color[i] and not bfs(i, 1, g, color):
            flag = False
            break
    if flag:
        print("Yes")
    else:
        print("No")
